Feat/1pr/media parser

polw1 opened the pull request:
MediaParser v0.1.8 - PR Description

Rust library for extracting thumbnails, subtitles, and metadata from MP4 files, with support for both local files and remote streaming.

### Key Features
- Thumbnails:  H.264 keyframe extraction with resizing
- Subtitles: Support for TX3G, WebVTT, TTML with SRT conversion
-  Metadata: Complete MP4 container analysis and iTunes metadata
- Streaming: Optimized HTTP range requests for remote files

### Architecture
- Modules `mp4/`, `avc/`, `thumbnails/`, `subtitles/`, `metadata/`, `streams/`


### Simple API
```rust
// Thumbnails
let thumbnails = extract_local_thumbnails("video.mp4", 5, 320, 240)?;

// Subtitles  
let subtitles = extract_local_subtitle_entries("video.mp4")?;

// Metadata
let metadata = read_local_complete_metadata("video.mp4")?;
```

polw1 committed:
feat: Initial release of MP4 parsing library

A high-performance Rust library for intelligent extraction of media information
from MP4 files, with support for both local and remote (HTTP) sources.

Core Features:
 * Efficient MP4 container parsing with minimal memory usage
 * Smart thumbnail extraction using I-frame detection
 * Subtitle track extraction with multiple format support
 * Comprehensive metadata extraction
 * HTTP range request optimization for remote files

Technical Implementation:
 * Uses seekable stream abstraction for unified local/remote handling
 * Implements efficient moov box parsing for quick metadata access
 * Employs H.264/AVC decoder for thumbnail generation
 * Supports various subtitle formats (TX3G, WebVTT, TTML)
 * Provides robust error handling with detailed messages

Design Principles:
 * Minimize memory usage through strategic data loading
 * Optimize network usage for remote files
 * Provide clear, strongly-typed APIs
 * Maintain comprehensive error handling
 * Support future format extensions

polw1 committed:
docs: Add MP4 format specification reference

Update MP4 format documentation to include the official ISO/IEC standard
reference number (14496-12) for better technical clarity and compliance
tracking.

velocitysystems commented:
@jthomerson See results of license audit below.

### Dependencies
1. **openh264**  
   - Crate: [openh264](https://crates.io/crates/openh264)  
   - License: MIT OR Apache-2.0  
   - Note: The underlying OpenH264 library from Cisco is licensed under BSD, but the Rust crate uses MIT/Apache-2.0.

2. **serde**  
   - Crate: [serde](https://crates.io/crates/serde)  
   - License: MIT OR Apache-2.0

3. **base64**  
   - Crate: [base64](https://crates.io/crates/base64)  
   - License: MIT OR Apache-2.0

4. **image**  
   - Crate: [image](https://crates.io/crates/image)  
   - License: MIT OR Apache-2.0

5. **reqwest**  
   - Crate: [reqwest](https://crates.io/crates/reqwest)  
   - License: MIT OR Apache-2.0

### Dev-Dependencies
1. **proptest**  
   - License: MIT OR Apache-2.0

2. **tempfile**  
   - License: MIT OR Apache-2.0

3. **wiremock**  
   - License: MIT OR Apache-2.0

4. **tokio**  
   - License: MIT OR Apache-2.0

5. **mockall**  
   - License: MIT OR Apache-2.0

6. **regex**  
   - License: MIT OR Apache-2.0

---

### Summary
- All dependencies and dev-dependencies use permissive licenses: **MIT OR Apache-2.0** (with the exception of openh264's underlying C library, which is BSD, but the crate is MIT/Apache-2.0).
- No copyleft or restrictive licenses are present.

polw1 committed:
refactor: Implement PR review suggestions for improved robustness

Major improvements to the codebase based on PR feedback to enhance
stability, safety, and maintainability.

Core Changes:
 * Update Rust edition to 2021 for better ecosystem compatibility
 * Replace println! with log crate for proper logging
 * Add 30s default HTTP timeout for remote operations
 * Implement 50MB size limit for moov box to prevent OOM
 * Unify public API with single method per function
 * Making funcionts async

Technical Improvements:
 * Add guard against extremely large moov boxes:
   - Implement MAX_MOOV_SIZE constant (50MB)
   - Add size validation before buffer allocation
   - Return proper error for oversized boxes

 * Streamline public API:
   - Unified extract_thumbnails method
   - Unified extract_subtitles method
   - Unified extract_metadata method
   - Automatic local/remote path detection

 * Enhance error handling:
   - Add proper error types
   - Improve error messages
   - Better error propagation

polw1 committed:
fix: Migrate stco.rs to use existing error system and improve type
flexibility

Replace io::Error usage in stco.rs with the library's existing
error types from errors.rs for better consistency.

Changes:
 * Replace io::Error with Mp4Error from errors.rs
 * Update error propagation in bit operations
 * Maintain consistent error handling across codebase

Type Flexibility:
 * Add support for both String and &str in function parameters
 * Maintain backward compatibility with existing code
 * Reduce unnecessary string allocations

polw1 committed:
refactor: Migrate bit reader to use existing error system

Replace io::Error usage in bit reader with the library's existing
error types from errors.rs for better consistency.

Changes:
 * Replace io::Error with BitReaderError from errors.rs
 * Update error propagation in bit operations
 * Maintain consistent error handling across codebase

polw1 committed:
chore: Set initial package version to 1.0.0

Update package version in Cargo.toml to follow semantic versioning
best practices for initial release.

Changes:
 * Update version from 0.1.9 to 1.0.0 in Cargo.toml
 * Establish baseline for future semantic versioning

velocitysystems commented on tests/testdata/big_buck_bunny.mp4:
Please update the README.md to indicate that the Big Buck Bunny video is licensed under the [Creative Commons Attribution 3.0 license](http://creativecommons.org/licenses/by/3.0/).

velocitysystems replied:
@polw1 Can you please rebase and update the commit messages so they conform to the Silvermine [standards](https://github.com/silvermine/silvermine-info/blob/master/commit-history.md)? Once this is done I'll hand you over to @yokuze for further review. Thanks!

polw1 committed:
docs: Add Big Buck Bunny license attribution

Add proper attribution for test video assets in README.md to comply
with licensing requirements.

Changes:
 * Add attribution for Big Buck Bunny test video
 * Include copyright information (© 2008, Blender Foundation)
 * Add Creative Commons Attribution 3.0 license reference

polw1 committed:
refactor: Remove unused main.rs file

Remove main.rs as this project is a library-only crate and does not
require an executable entry point.

Changes:
 * Delete src/main.rs file

yokuze commented on Cargo.toml:
I'm concerned by the large amount of code we have here for parsing the MP4 format that we will have to maintain, but I also don't have all of the context since I think I was away when this library was specced out/ discussed. 

Do we need to write our own MP4 parser, or is there a Rust crate that could help us out with that so this lib can just provide a convenient interface for getting the metadata/thumbnails/etc. for media files without being directly responsible for the complex business of parsing?

Similarly, could we benefit from existing crates that parse the various subtitle formats?

@velocitysystems Need your input here.

yokuze commented on src/thumbnails/doc.md:
````diff
@@ -0,0 +1,210 @@
+```rust
+use mediaparser::thumbnails::extract_thumbnails;
+
+let thumbnails = extract_thumbnails("video.mp4", 5, 320, 240).await;
````
Should we allow for specifying what video track to extract the thumbnails from? It looks like we default to the first one we find. Probably the most flexible API would be allowing the caller to pass in a callback function that takes in a list of available tracks and returns the index of the track to use for thumbnails.

Perhaps that could be an enhancement for another PR. If we switch to accepting a config object rather than individual parameters, it'd be easy to add a new property that supports that.

yokuze commented on src/subtitles/types.rs:
```diff
@@ -0,0 +1,30 @@
+
+#[derive(Debug)]
+pub(crate) struct SubtitleTrackInfo {
+    pub _track_id: u32,
```
Why does this public property start with an underscore?

yokuze commented on src/subtitles/types.rs:
```diff
@@ -0,0 +1,30 @@
+pub(crate) struct SubtitleSampleRange {
+    pub offset: u64,
+    pub size: u32,
+    pub _sample_index: u32,
```
Why does this public property start with an underscore?

yokuze commented on src/subtitles/extractor.rs:
```diff
@@ -0,0 +1,188 @@
+}
+
+/// Calculate sample timestamps from timing tables
+#[allow(dead_code)]
```
Why? We don't generally allow unused or commented-out code.

velocitysystems replied:
Agree. Any dead code should be removed.

yokuze commented on src/subtitles/doc.md:
````diff
@@ -0,0 +1,206 @@
+### SubtitleEntry Structure
+```rust
+pub struct SubtitleEntry {
+    pub start: String,    // SRT format: "00:00:05,000"
````
Are the timestamp formats going to vary across subtitle formats? If so, should we either:

1. Normalize to a common format
2. Include the subtitle type in `extract_subtitles`'s return value

Probably it would nice to include #2 in any case

yokuze commented on src/subtitles/doc.md:
````diff
@@ -0,0 +1,206 @@
+```rust
+use mediaparser::extract_subtitles;
+
+let entries = extract_subtitles("video.mp4").await;
````
We need to support files with multiple subtitle tracks in different languages, and the data returned should include the language locale if that information is available.

I'm also not sure what our immediate use case will be for these extracted subtitles-- I wasn't there for that discussion. Search indexing? If you know @velocitysystems could you review this API and make sure it's returning all of the data we would need?

yokuze commented on src/subtitles/parser.rs:
```diff
@@ -0,0 +1,205 @@
+        if !text.trim().is_empty() {
+            return Ok(vec![SubtitleEntry {
+                start: format_timestamp(timestamp),
+                end: format_timestamp(timestamp + 2.0), // Default 2-second duration
```
I may be misunderstanding what's happening, but I don't think that assuming that all subtitles have 2 second duration is what we want. We should be grabbing the real end times specified in the subtitle data, right?

This doesn't seem to be a default as much as it is just overriding all of the end timestamp data.

yokuze reviewed (changes requested):
@polw1 Thank you very much for all of your hard work on this crate! It's exciting to see some working code samples that extracts all of this data for us.

I've left some comments and questions for you. I grouped several together just on the Cargo.toml file because addressing them affects many different files.

When you are addressing MR comments, please follow the Silvermine git standards outlined here:

https://github.com/silvermine/silvermine-info/blob/master/commit-history.md

and fixup/squash your "fix/refactor" commits back into the commits that introduced the code. For example:

<img width="578" height="85" alt="image" src="https://github.com/user-attachments/assets/972d6b6f-5123-4824-b57d-9cf3f5ed993c" />

^^ the changes in this commit should be squashed into whatever commit(s) introduced the code (as long as the commit is also in this PR)

Please let me know if you have any questions on my comments or want to collaborate on API design, etc.

yokuze commented:
An idea for future large projects like this: It can lead to better quality reviews and outcomes if the PRs come in phases. For example, on a recent project we started with just the public documentation for the library and the types/interfaces. The first PR was entirely markdown documentation and a few types/interfaces, and that helped us collaborate on what the API would look like, and make sure we were covering all of the use cases, etc.

But I've also done the whole "massive initial PR" thing several times, so no judgement at all here 😅 Just a suggestion that's worked out well in the past.

velocitysystems commented on README.md:
```diff
@@ -1,6 +1,131 @@
+# MediaParser
 
-## What?
+A high-performance Rust library for intelligent extraction of thumbnails, subtitles, and metadata from MP4 video files. Designed for both local and remote file processing with optimized streaming capabilities.
```
Suggestion: "A Rust library for the extraction of thumbnails, subtitles and metadata from MP4 video files. Supports both local and remote video files."

velocitysystems commented on src/avc/annexb.rs:
```diff
@@ -0,0 +1,133 @@
+
+// Legacy functions for backward compatibility
+/// Extract the first video NAL unit from a bytestream.
+#[deprecated(
```
Talking with @polw1 we _may_ use this in the future but we can remove it from this PR.

velocitysystems commented on tests/lib.rs:
```diff
@@ -0,0 +1 @@
+
```
Let's remove this for now.

jthomerson reviewed (changes requested):
Hey guys, I wonder if we should take a step back and re-assess how this initial MVP of the library lands? Rather than one PR that's so large that it will take forever to review and land, could we re-assess our strategy here? There's a lot of questions in this PR about even the public interface that the library exposes, which is the main building block of building a good library. Should we pivot to a new PR that focuses solely on defining a no-op API that is agreed upon? (we may need one PR prior to that which just sets up project structure, README, linting, etc). That API-only PR basically then serves as our requirements-gathering stage - we define the API we need, and then land it, even though it has no implementation. Then we follow up with subsequent PR's to build the individual pieces defined in those requirements.

As @yokuze mentions, we also need to determine whether we really need to write our own MP4 parsing or if we can use crates already out there for individual pieces. Maybe it's best to write our own - I don't know. But we should at least discuss it, maybe in an issue, and document why we went one way or the other.

Thanks for considering all this feedback.
